### PR TITLE
Issue number - 519 : Introduce validation logic for elements in configuration file

### PR DESCRIPTION
**Pull request by chaktatha**
…ser and ExtendedConfigParser

### CODE REVIEW

File: lease/entity-server/src/test/java/org/terracotta/lease/service/config/LeaseConfigValidatorTest.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+
+public class LeaseConfigValidatorTest {
+
+  @Test
+  public void testValidate() throws ValidationException {
```
Comment (ramsai1729) [inline lease/entity-server/src/test/java/org/terracotta/lease/service/config/LeaseConfigValidatorTest.java]: Add one more test with different time units but with same lease time?

File: lease/entity-server/src/main/java/org/terracotta/lease/service/config/LeaseConfigurationParser.java [hunk 2]
```diff
@@ -102,4 +96,33 @@ public LeaseConfiguration parse(Element element, String source) {
...
+      Element leaseLengthElement = (Element)childElements.item(0);
+
+      String leaseLengthString = leaseLengthElement.getTextContent();
+      LOGGER.info("Found lease length XML text: " + leaseLengthString);
+
+      String timeUnitString = leaseLengthElement.getAttribute(TIME_UNIT_ATTRIBUTE_NAME);
+      LOGGER.info("Found lease length time unit: " + timeUnitString);
+
+      LeaseElement leaseElement = new LeaseElement();
+      leaseElement.setLeaseValue(leaseLengthString);
```
Comment (ramsai1729) [inline lease/entity-server/src/main/java/org/terracotta/lease/service/config/LeaseConfigurationParser.java]: can't we set these values in the constructor?

File: lease/entity-server/src/main/java/org/terracotta/lease/service/config/LeaseConfigurationParser.java [hunk 3]
```diff
@@ -102,4 +96,33 @@ public LeaseConfiguration parse(Element element, String source) {
 
     return new LeaseConfiguration(TimeUnit.MILLISECONDS.convert(leaseLength, timeUnit));
   }
-}
+
+  public Function<Element, LeaseElement> parser() {
```
Comment (ramsai1729) [inline lease/entity-server/src/main/java/org/terracotta/lease/service/config/LeaseConfigurationParser.java]: private?

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapConfigValidator.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.stream.Collectors;
+
+public class OffHeapConfigValidator implements ConfigValidator {
+
+  private final Map<MemoryUnit, Long> conversionMap = new HashMap<>();
+  private final Function<Element, OffheapResourcesType> resourcesTypeSupplier;
+
+  public OffHeapConfigValidator(Function<Element, OffheapResourcesType> resourcesTypeSupplier) {
+    conversionMap.put(MemoryUnit.B, 1L);
+    conversionMap.put(MemoryUnit.K_B, 1024L);
```
Comment (ramsai1729) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapConfigValidator.java]: this should be 'kB', right?

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapConfigValidator.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class OffHeapConfigValidator implements ConfigValidator {
+
+  private final Map<MemoryUnit, Long> conversionMap = new HashMap<>();
+  private final Function<Element, OffheapResourcesType> resourcesTypeSupplier;
+
+  public OffHeapConfigValidator(Function<Element, OffheapResourcesType> resourcesTypeSupplier) {
+    conversionMap.put(MemoryUnit.B, 1L);
```
Comment (ramsai1729) [inline offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapConfigValidator.java]: can we reuse the conversions utilities in OffHeapResourcesProvider

File: lease/entity-server/src/test/java/org/terracotta/lease/service/config/LeaseConfigValidatorTest.java (hunk 1 above)
Comment (chaktatha) [reply lease/entity-server/src/test/java/org/terracotta/lease/service/config/LeaseConfigValidatorTest.java]: testValidateWithMismatchedLeaseTimeUnit is for that purpose

File: lease/entity-server/src/main/java/org/terracotta/lease/service/config/LeaseConfigurationParser.java (hunk 2 above)
Comment (chaktatha) [reply lease/entity-server/src/main/java/org/terracotta/lease/service/config/LeaseConfigurationParser.java]: Will do

File: lease/entity-server/src/main/java/org/terracotta/lease/service/config/LeaseConfigurationParser.java (hunk 3 above)
Comment (chaktatha) [reply lease/entity-server/src/main/java/org/terracotta/lease/service/config/LeaseConfigurationParser.java]: Will make the change

File: offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapConfigValidator.java (hunk 5 above)
Comment (chaktatha) [reply offheap-resource/src/main/java/org/terracotta/offheapresource/OffHeapConfigValidator.java]: Ok. Can be done

File: lease/entity-server/src/test/java/org/terracotta/lease/service/config/LeaseConfigValidatorTest.java (hunk 1 above)
Comment (ramsai1729) [reply lease/entity-server/src/test/java/org/terracotta/lease/service/config/LeaseConfigValidatorTest.java]: meant one element with 2 minutes and another element with 120 seconds for e.g.

Comment (chaktatha) [reply lease/entity-server/src/test/java/org/terracotta/lease/service/config/LeaseConfigValidatorTest.java]: Oh ok. I will do.

**Pull request merged**